### PR TITLE
Add dex expect type command

### PR DESCRIPTION
**Pull request by TomWeaver18**
You can now write tests that check the type of a variable or expression at some point during test execution.

The heuristic scoring works exactly the same way as DexExpectWatchValue, and as a result, I've now implemented a DexExpectWatchBase command that Value and Type checking both inherit from.

There's a small change to LLDB, for type information there's now an extra step that attempts to find expressions using the variable finding API. The FindVariable() method returns more accurate type information for expressions that evaluate to variables with distinct locations. If this fails we simply use the original expression type information.

Docs have been updated and I've also added two feature tests, one for perfect and penalty.

StepValueInfo has now been moved to a seperate file in the commands folder. We may want to move it somewhere else, I welcome discussion on this.

few stealth changes for misformatted comments and added @staticmethod properties where they were missing.

happy reviewing.

### CODE REVIEW

File: feature_tests/commands/penalty/expect_watch_type/test.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Purpose:
+//      Check that \DexExpectWatchType applies no penalties when expected
```
Comment (OCHyams) [inline feature_tests/commands/penalty/expect_watch_type/test.cpp]: This description isn't correct (copy pasta job?).

File: dex/command/commands/DexExpectWatchBase.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    @staticmethod
+    @abc.abstractstaticmethod
+    def get_name():
+        """This abstract method is usually implemented in subclasses as:
+        return __class__.__name__
+        """
+
+    @abc.abstractmethod
+    def _expected_field(self, watch):
+        """Return attribute from watch that we're most interested in.
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchBase.py]: the word "value" is overloaded here as we have DexExpectWatchValue for checking expression values and DexExpectWatchType for checking expression/variable types. I think "_get_expected_field" would be better, but maybe not the best we can come up with.

File: dex/debugger/lldb/LLDB.py [hunk 3]
```diff
@@ -236,10 +237,19 @@ def evaluate_expression(self, expression, frame_idx=0) -> ValueIR:
         if error_string == 'success':
             error_string = None
 
+        # attempt to find expression as a variable, if found, take the variable
+        # obj's type information as it's 'usually' more accurate.
+        var_result = self._thread.GetFrameAtIndex(frame_idx).FindVariable(expression)
+        type_name = str()
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/debugger/lldb/LLDB.py]: this is a definition of type_name for use later in the type assignment of the ValueIR returned by the debugger. type_name becomes an alias for the type information we conditionally select shortly after.

unless you dont need to forward declare the variable as type str()?

File: dex/command/commands/DexExpectWatchValue.py [hunk 4]
```diff
@@ -22,65 +22,10 @@
...
-            index += 1
-        else:
-            assert False, 'unexpected diff:{}'.format(vd)
-
-    return differences
-
-
-class DexExpectWatchValue(CommandBase):
+class DexExpectWatchValue(DexExpectWatchBase):
     """Expect the C++ expression `expr` to evaluate to the list of `values`
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchValue.py]: will do, thanks for the spot

File: feature_tests/commands/penalty/expect_watch_type/test.cpp (hunk 1 above)
Comment (TomWeaver18) [reply feature_tests/commands/penalty/expect_watch_type/test.cpp]: thanks for the spot, will sort this .

File: dex/command/commands/DexExpectWatchBase.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+"""ExpectWatch base class, holds logic for how to build and process expected
+ watche commands.
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchBase.py]: Thanks for the spot, will get this fixed.

File: dex/debugger/lldb/LLDB.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply dex/debugger/lldb/LLDB.py]: `if` blocks (or, any control flow) in python do not introduce a nested scope like in C++ so you don't need to declare it beforehand.

File: dex/command/commands/DexExpectWatchBase.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply dex/command/commands/DexExpectWatchBase.py]: I agree that the word 'value' is overloaded. As long as the docstring is updated and explains what the fn is for we can live with a slightly off name.

File: Commands.md [hunk 6]
```diff
@@ -83,13 +83,40 @@ Expect the line every `DexExpectStepOrder` is found on to be stepped on in
 [TODO]
 
 
+---
+## DexExpectWatchValue
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply Commands.md]: will do, thanks for the spot.

File: dex/debugger/lldb/LLDB.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/debugger/lldb/LLDB.py]: righto, I'll get that whipped out. thanks for the spot!

File: dex/command/commands/DexExpectWatchBase.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    def has_labels(self):
+        return len(self.get_label_args()) > 0
+
+    def get_label_args(self):
+        return [label for label in (self._from_line, self._to_line)
+                      if isinstance(label, str)]
+
+    @staticmethod
+    @abc.abstractstaticmethod
+    def get_name():
```
[the remark this answers is not in the document]
Comment (TomWeaver18) [reply dex/command/commands/DexExpectWatchBase.py]: removed the whole things, thanks for the headsup on this.

**Comment by TomWeaver18** [comment]
force pushed with requested changes.

**Comment by OCHyams** [comment]
LGTM. Should get another set of eyes on this since it's large-ish though.

**Comment by TomWeaver18** [comment]
👍 Cheers bud!

**Pull request merged**